### PR TITLE
Enable VPP output on [non]linear

### LINKED ISSUE #24705: Support VectorPostprocessor output on `NONLINEAR` and `LINEAR`
## Reason
For debugging purposes it can be very useful to examine VPP results at steps other than `TIMESTEP_END`. Currently this is not supported, simply because the VPP filenames only contain the time step (and multiple outputs during the same time step would repeatedly overwrite the same file).

## Design
Append the nonlinear (and if needed linear) iteration numbers to the filename.

## Impact
Added capability.

### PR DESCRIPTION
**Pull request by dschwen**
Closes #24705 

(still need to add tests)

### CODE REVIEW

**Comment by GiudGiud** [comment]
😄

**Comment by GiudGiud** [comment]
ah you need to unsubscribe manually too

**Comment by dschwen** [comment]
I have PRs ready for subchannel (pronghorn) and griffin.

**Comment by dschwen** [comment]
One key change in this PR is that not every `PetscOutput` object registers its own monitor. I actually ran into a PETSc error where too many monitors were registered, if you perform debugging output on linear and nonlinear iterations with multiple outputs.

File: framework/include/outputs/PetscOutput.h [hunk 1]
```diff
@@ -101,4 +121,6 @@ class PetscOutput : public Output
 
   /// Linear residual output end time
   Real _linear_end_time;
+
+  friend class PetscOutputInterface;
```
Comment (GiudGiud) [inline framework/include/outputs/PetscOutput.h:125]: why do you need to be friend with a class you inherit

File: framework/src/outputs/PetscOutput.C [hunk 2]
```diff
@@ -249,3 +260,16 @@ PetscOutput::time()
   else
     return Output::time();
 }
+
+bool
+PetscOutput::shouldOutput()
+{
+  if (Output::shouldOutput())
+  {
+    if (_current_execute_flag == EXEC_NONLINEAR &&
```
Comment (GiudGiud) [inline framework/src/outputs/PetscOutput.C:269]: document that logic and say why this works for linear too

Comment (GiudGiud) [reply framework/src/outputs/PetscOutput.C:269]: also can you use inNonlinearTimeWindow instead?

**Review by GiudGiud — COMMENTED** [review]
some minor comments

File: framework/include/outputs/PetscOutput.h (hunk 1 above)
Comment (dschwen) [reply framework/include/outputs/PetscOutput.h:125]: Other direction

**Review by GiudGiud — APPROVED** [review]

**Comment by dschwen** [comment]
Rebasing and pushing once more to et a fresh set of tests. Griffin PR has been ready for a while. I'd like to get this in this week.

**Comment by GiudGiud** [comment]
weird that my approval did not get dismissed by your last push

**Comment by GiudGiud** [comment]
my guess is that rebase-only dont dismiss approval

**Review by GiudGiud — APPROVED** [review]

**Pull request merged**